buildpack/descriptor.go

```diff
@@ -15,14 +40,56 @@ func (b *Descriptor) ConfigFile() *Descriptor {
 	return b
 }
 
-func (b *Descriptor) IsMetaBuildpack() bool {
-	return b.Order != nil
+func (b *Descriptor) Info() *Info {
+	switch {
+	case b.IsBuildpack():
+		return &b.Buildpack
+	case b.IsExtension():
+		return &b.Extension
+	}
+	return &Info{}
+}
+
+func (b *Descriptor) IsBuildpack() bool {
+	return b.Buildpack.ID != ""
+}
+
+func (b *Descriptor) IsComposite() bool {
+	return len(b.Order) > 0
+}
+
+func (b *Descriptor) IsExtension() bool {
+	return b.Extension.ID != ""
+}
+
+func (b *Descriptor) Kind() string {
+	if b.IsExtension() {
+		return "extension"
+	}
+	return "buildpack"
 }
 
 func (b *Descriptor) String() string {
 	return b.Buildpack.Name + " " + b.Buildpack.Version
 }
 
+func (b *Descriptor) ToGroupElement() GroupElement {
+	groupEl := GroupElement{API: b.API}
+	switch {
+	case b.IsBuildpack():
+		groupEl.ID = b.Buildpack.ID
+		groupEl.Version = b.Buildpack.Version
+		groupEl.Homepage = b.Buildpack.Homepage
+	case b.IsExtension():
+		groupEl.ID = b.Extension.ID
+		groupEl.Version = b.Extension.Version
+		groupEl.Homepage = b.Extension.Homepage
+		groupEl.Extension = true
+		groupEl.Optional = true
```

Comment:
You are right, it's a bit weird. TBH at the point where `ToGroupElement()` is used, we don't really care if the element is optional or not - this feeds through to the output of "plan resolution" which is passed from the detect phase to the build phase, and the build phase doesn't have a concept of optional / not optional for buildpacks, it just runs what it is given. We could just not set this for extensions, but since extensions are always optional it doesn't hurt. I think what would ultimately make this better is to pull apart `GroupElement` and have a separate struct for each of the scenarios it currently covers (see https://github.com/buildpacks/lifecycle/pull/860#discussion_r880889983) but that is a pretty big refactor to tackle along with all of these other changes 😐

Reply:
Removed `ToGroupElement()` weirdness in bc0a6460.